[BUG] regex escapes for responses not correctly validated

### Checklist

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)
- [x] I am using the latest version of Schemathesis

### Describe the bug

I have a schema that contains a regex for a docker image name, which uses escapes like `\\/` in the pattern. Since version 3.39.8, this fails tests with returned value not matching the pattern. This is likely due to the escape sequences not being handled correctly.

### To Reproduce

Please include a minimal API schema causing this issue:

the component schema as returned on the spec.json endpoint
```json
"ContainerImage": {
    "description": "A container image",
    "type": "string",
    "maxLength": 500,
    "pattern": "^[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*(\\/[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*)*(:[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}|@sha256:[a-fA-F0-9]{64}){0,1}$",
    "example": "renku/renkulab-py:3.10-0.18.1"
  }
```

Note the example is a value that will fail the schemathesis test (due to the `/`).

[schemathesis test this happens on](https://github.com/SwissDataScienceCenter/renku-data-services/blob/main/test/bases/renku_data_services/data_api/test_schemathesis.py#L91)

Example error log:
```
cls = <class 'jsonschema.validators.Draft4Validator'>
schema = {'description': 'A container image', 'example': 'renku/renkulab-py:3.10-0.18.1', 'pattern': '^([a-z0-9]){1,500}(\\.|_|...z0-9]+((\\.|_|__|-+)[a-z0-9]+)*){0}(:[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}|@sha256:[a-fA-F0-9]{64}){0}$', 'type': 'string'}
format_checker = <FormatChecker checkers=['date-time', 'email', 'hostname', 'idn-email', 'ipv4', 'ipv6', 'regex', 'uri']>

    @classmethod
    def check_schema(cls, schema, format_checker=_UNSET):
        Validator = validator_for(cls.META_SCHEMA, default=cls)
        if format_checker is _UNSET:
            format_checker = Validator.FORMAT_CHECKER
        validator = Validator(
            schema=cls.META_SCHEMA,
            format_checker=format_checker,
        )
        for error in validator.iter_errors(schema):
>           raise exceptions.SchemaError.create_from(error)
E           jsonschema.exceptions.SchemaError: '^([a-z0-9]){1,500}(\\.|_|__|-+)[a-z0-9]+){0}(\\/[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*){0}(:[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}|@sha256:[a-fA-F0-9]{64}){0}$' is not a 'regex'
E           
E           Failed validating 'format' in metaschema['properties']['pattern']:
E               {'type': 'string', 'format': 'regex'}
E           
E           On schema['pattern']:
E               '^([a-z0-9]){1,500}(\\.|_|__|-+)[a-z0-9]+){0}(\\/[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*){0}(:[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}|@sha256:[a-fA-F0-9]{64}){0}$'
E           while generating 'case' from one_of(get_case_strategy(operation=APIOperation(path='/session_launchers/{launcher_id}',
E            method='patch',
E            definition=,
E            schema=,
E            verbose_name='PATCH /api/data/session_launchers/{launcher_id}',
E            app=None,
E            base_url='/api/data',
E            path_parameters=,
E            headers=,
E            cookies=,
E            query=,
E            body=,
E            case_cls=), generation_config=GenerationConfig(allow_x00=True,
E            graphql_allow_null=True,
E            codec='utf-8',
E            with_security_parameters=True,
E            headers=HeaderConfig(strategy=None)), skip_on_not_negated=False), get_case_strategy(operation=APIOperation(path='/session_launchers/{launcher_id}',
E            method='patch',
E            definition=,
E            schema=,
E            verbose_name='PATCH /api/data/session_launchers/{launcher_id}',
E            app=None,
E            base_url='/api/data',
E            path_parameters=,
E            headers=,
E            cookies=,
E            query=,
E            body=,
E            case_cls=), generator=<DataGenerationMethod.negative: 'negative'>, generation_config=GenerationConfig(allow_x00=True,
E            graphql_allow_null=True,
E            codec='utf-8',
E            with_security_parameters=True,
E            headers=HeaderConfig(strategy=None)), skip_on_not_negated=False))
```

### Expected behavior

Test should not fail

### Environment

```
- OS: Linux
- Python version: 3.12
- Schemathesis version: 3.39.8+
- Spec version: Open API 3.0.2
```

I tested with `3.39.12` and unfortunately it still fails with
```
E       schemathesis.exceptions.CheckFailed:
E
E       1. Response violates schema
E
E           'renku/renkulab-py:3.10-0.18.1' does not match '^([a-z0-9]){1,500}((\\.|_|__|-+)[a-z0-9]+){0}(\\/[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*){0}(:[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}|@sha256:[a-fA-F0-9]{64}){0}$'
E
E           Schema at /properties/container_image:
E
E               {
E                   "description": "A container image",
E                   "type": "string",
E                   "pattern": "^([a-z0-9]){1,500}((\\.|_|__|-+)[a-z0-9]+){0}(\\/[a-z0-9]+((\...
E                   "example": "renku/renkulab-py:3.10-0.18.1"
E               }
E
E           Value:
E
E               "renku/renkulab-py:3.10-0.18.1"
E
E       [201] Created:
E
E           `{"id":"01JNJV22Q3DBXP7VPK5E0KVG7Z","name":"My Renku Session :)","creation_date":"2025-03-05T09:44:16Z","container_image":"renku/renkulab-py:3.10-0.18.1","default_url":"/lab","uid":1000,"gid":1000,"working_directory":"/home/jovyan/work","mount_directory":"/home/jovyan/work","port":8080,"is_archived":false}`
E
E       Reproduce with:
E
E           curl -X POST -H 'host: mockserver:1234' -H 'authorization: Bearer {"is_admin": true, "id": "admin", "name": "Admin Doe", "first_name": "Admin", "last_name": "Doe", "email": "admin.doe@gmail.com", "full_name": "Admin Doe"}' -H 'content-type: application/json' -d '{"name": "My Renku Session :)", "container_image": "renku/renkulab-py:3.10-0.18.1", "default_url": "/lab", "working_directory": "/home/jovyan/work", "mount_directory": "/home/jovyan/work"}' http://localhost/api/data/environments
E
E       Falsifying explicit example: async_run(
E           sanic_client=<test.utils.SanicReusableASGITestClient object at 0x7402717c03b0>,
E           admin_headers={'Authorization': 'Bearer {"is_admin": true, "id": "admin", "name": "Admin Doe", "first_name": "Admin", "last_name": "Doe", "email": "admin.doe@gmail.com", "full_name": "Admin Doe"}'},
E           requests_statistics=RequestsStatistics(len=465),
E           case=,
E       )
```

what's odd to me is, in the error you got, the part at the beginning of the regex `^([a-z0-9]){1,500}(\.|_|__|-+)[a-z0-9]+){0}` which has unbalanced parenthesis but if I access the link to our spec, it looks like `"pattern":"^[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*(\\/[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*)*(:[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}|@sha256:[a-fA-F0-9]{64}){0,1}$"` so there's something other than just escaping going on it seems? specifically, the {1,500} is not in the pattern on our side, though I assume that was added due to the min and max length constraints? and it seems that instead of a `*` in your error message there's a `{0}`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,10 @@
 
 - **INTERNAL**: Use `str` in enums.
 
+### :bug: Fixed
+
+- Do not mutate `pattern` keywords in response schema validation. [#2749](https://github.com/schemathesis/schemathesis/issues/2749)
+
 ## [4.0.0-alpha.7](https://github.com/schemathesis/schemathesis/compare/v4.0.0-alpha.6...v4.0.0-alpha.7) - 2025-03-21
 
 ### :rocket: Added

--- a/src/schemathesis/specs/openapi/references.py
+++ b/src/schemathesis/specs/openapi/references.py
@@ -139,7 +139,10 @@ class ConvertingResolver(InliningResolver):
     def resolve(self, ref: str) -> tuple[str, Any]:
         url, document = super().resolve(ref)
         document = to_json_schema_recursive(
-            document, nullable_name=self.nullable_name, is_response_schema=self.is_response_schema
+            document,
+            nullable_name=self.nullable_name,
+            is_response_schema=self.is_response_schema,
+            update_quantifiers=False,
         )
         return url, document
 

--- a/test/specs/openapi/test_patterns.py
+++ b/test/specs/openapi/test_patterns.py
@@ -2,6 +2,7 @@ import re
 import sys
 
 import pytest
+from flask import Flask, jsonify
 from hypothesis import HealthCheck, assume, given, settings
 from hypothesis import strategies as st
 
@@ -170,3 +171,71 @@ def is_valid_regex(pattern: str) -> bool:
         return True
     except re.error:
         return False
+
+
+def test_response_schema_is_not_mutated(cli, app_runner, snapshot_cli):
+    # See GH-2749
+    raw_schema = {
+        "openapi": "3.0.3",
+        "info": {"title": "Container Image API", "version": "1.0.0"},
+        "paths": {
+            "/container": {
+                "post": {
+                    "summary": "Create a container image",
+                    "requestBody": {
+                        "required": True,
+                        "content": {
+                            "application/json": {
+                                "schema": {
+                                    "type": "object",
+                                    "properties": {"container_image": {"$ref": "#/components/schemas/ContainerImage"}},
+                                }
+                            }
+                        },
+                    },
+                    "responses": {
+                        "200": {
+                            "description": "Successful response",
+                            "content": {
+                                "application/json": {
+                                    "schema": {
+                                        "type": "object",
+                                        "properties": {
+                                            "container_image": {"$ref": "#/components/schemas/ContainerImage"}
+                                        },
+                                    }
+                                }
+                            },
+                        }
+                    },
+                }
+            }
+        },
+        "components": {
+            "schemas": {
+                "ContainerImage": {
+                    "description": "A container image",
+                    "type": "string",
+                    "maxLength": 500,
+                    "pattern": "^[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*(\\/[a-z0-9]+((\\.|_|__|-+)[a-z0-9]+)*)*(:[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}|@sha256:[a-fA-F0-9]{64}){0,1}$",
+                    "example": "renku/renkulab-py:3.10-0.18.1",
+                }
+            }
+        },
+    }
+
+    app = Flask(__name__)
+
+    @app.route("/openapi.json")
+    def openapi_spec():
+        return jsonify(raw_schema)
+
+    @app.route("/container", methods=["POST"])
+    def create_container():
+        example_value = raw_schema["components"]["schemas"]["ContainerImage"]["example"]
+        response_body = {"container_image": example_value}
+        return jsonify(response_body), 200
+
+    port = app_runner.run_flask_app(app)
+
+    assert cli.run(f"http://127.0.0.1:{port}/openapi.json", "-call", "--phases=fuzzing", "-n 1") == snapshot_cli

